lib/improver/blending/calculate_weights_and_blend.py

```diff
@@ -0,0 +1,227 @@
+                cube, self.blend_coord)
+
+        return weights
+
+    def _update_spatial_weights(self, cube, weights, fuzzy_length):
+        """
+        Update weights using spatial information
+
+        Args:
+            cube (iris.cube.Cube):
+                Cube of input data to be blended
+            weights (iris.cube.Cube):
+                Initial 1D cube of weights scaled by self.weighting_coord
+            fuzzy_length (float):
+                Distance (in metres) over which to smooth weights at domain
+                boundaries
+
+        Returns:
+            weights (iris.cube.Cube):
+                Updated 3D cube of spatially-varying weights
+        """
+        check_if_grid_is_equal_area(cube)
+        grid_cells_x, _ = convert_distance_into_number_of_grid_cells(
+            cube, fuzzy_length, int_grid_cells=False)
+        SpatialWeightsPlugin = SpatiallyVaryingWeightsFromMask(grid_cells_x)
+        weights = SpatialWeightsPlugin.process(cube, weights, self.blend_coord)
+        return weights
+
+    def process(self, cubelist, weighting_mode='weighted_mean',
+                cycletime=None, model_id_attr=None,
+                spatial_weights=False, fuzzy_length=20000):
+        """
+        Merge a cubelist, calculate appropriate blend weights and blend into
+        a single cube over the self.blend_coord dimension.
+
+        Args:
+            cubelist (iris.cube.CubeList):
+                List of cubes to be merged and blended
+
+        Kwargs:
+            weighting_mode (str):
+                "weighted_mean" or "weighted_maximum"
+            cycletime (str):
+                Forecast reference time to use for output cubes, in the format
+                YYYYMMDDTHHMMZ.  If not set, the latest of the input cube
+                forecast reference times is used.
+            model_id_attr (str):
+                Name of the attribute by which to identify the source model and
+                construct "model" coordinates for blending.
+            spatial_weights (bool):
+                If true, calculate spatial weights.
+            fuzzy_length (float):
+                Distance (in metres) over which to smooth spatial weights.
+                Default is 20 km.
+        """
+        # prepare cubes for weighted blending, including creating model_id and
+        # model_configuration coordinates for multi-model blending.  Raises an
+        # error if blend_coord is not present on all input cubes.
+        merger = MergeCubesForWeightedBlending(
+            self.blend_coord, weighting_coord=self.weighting_coord,
+            model_id_attr=model_id_attr)
+        cube = merger.process(cubelist, cycletime=cycletime)
+
+        # if blend_coord has only one value, or is not present (case where only
+        # one model has been provided for a model blend) update metadata only
+        coord_names = [coord.name() for coord in cube.coords()]
+        if (self.blend_coord not in coord_names or
```

Comment:
You've got one bit of metadata updating (of attributes) in the MergeCubesForWeightedBlending plugin, but another bit (cycle times) here. Should they be in the same place? (At a first read through I was expected them to be together.)

Reply:
I also don't think you have a unit test which uses this code.

Reply:
Ideally, all of the metadata would be dealt with in the blending or top level plugin (ie by `conform_metadata`).  The reason the `blend` attribute is dealt with separately is because of how merging deals with mismatched attributes.

It would probably be possible at this point (now that merging has been refactored) to factor that out into `conform_metadata`.  However it would involve passing around an additional attribute name and possibly a little more fiddling with code.  So if you'd like this change I'd rather look into that as a separate PR.

Re unit tests, no I don't have anything that calls `conform_metadata` (or that `conform_metadata` would change).  I'll add something using the `cycletime` argument.

Reply:
Thanks for the unit test.

I'm quite happy for nothing to change. The only thing you could change is moving the cycletime updating into the merging functionality, as that's where the cycletime updating would occur for the case where you pass in multiple cubes and where the attributes are updated, but there may be reasons not to do that.